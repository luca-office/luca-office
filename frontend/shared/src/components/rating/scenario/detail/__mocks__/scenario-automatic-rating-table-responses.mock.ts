import {omit} from "lodash-es"
import {
  automatedCodingCriterionEvaluationResultMock,
  automatedCodingItemMock,
  scenarioAutomatedCodingItemRatingsMock,
  scenariosMock
} from "../../../../../graphql/__mocks__"
import {ScenarioRatingCriterionSelectionCreation} from "../../../../../graphql/generated/globalTypes"
import {
  createScenarioRatingCriterionSelectionMutation,
  deleteScenarioRatingCriterionSelectionMutation,
  updateScenarioCodingItemRatingMutation
} from "../../../../../graphql/mutations"
import {evaluationResultsForAutomatedCodingItemQuery} from "../../../../../graphql/queries"
import {ScenarioRatingCriterionSelection} from "../../../../../models"
import {MockedResponse} from "@apollo/client/testing"

const scenario = scenariosMock[0]
const scenarioRating = {...scenarioAutomatedCodingItemRatingsMock[0], scenarioId: scenario.id}

export const scenarioAutomaticRatingTableResponsesMock: MockedResponse[] = [
  {
    request: {
      query: createScenarioRatingCriterionSelectionMutation,
      variables: {
        creation: omit<
          ScenarioRatingCriterionSelection,
          keyof Omit<ScenarioRatingCriterionSelection, keyof ScenarioRatingCriterionSelectionCreation>
        >(scenarioRating.criterionSelections[0], ["__typename", "createdAt"])
      }
    },
    result: {
      data: {
        createScenarioRatingCriterionSelection: scenarioRating.criterionSelections[0]
      }
    }
  },
  {
    request: {
      query: deleteScenarioRatingCriterionSelectionMutation,
      variables: {
        scenarioRatingId: scenarioRating.id,
        automatedCriterionId: scenarioRating.criterionSelections[0].automatedCriterionId
      }
    },
    result: {
      data: {
        deleteScenarioRatingCriterionSelection: scenarioRating.criterionSelections[0]
      }
    }
  },
  {
    request: {
      query: evaluationResultsForAutomatedCodingItemQuery,
      variables: {
        id: automatedCodingItemMock.id,
        surveyInvitationId: scenarioRating.surveyInvitationId,
        scenarioId: scenarioRating.scenarioId
      }
    },
    result: {
      data: {
        evaluationResultsForAutomatedCodingItem: automatedCodingCriterionEvaluationResultMock
      }
    }
  },
  {
    request: {
      query: updateScenarioCodingItemRatingMutation,
      variables: {
        id: scenarioAutomatedCodingItemRatingsMock[0].id,
        update: {noCriterionFulfilled: scenarioAutomatedCodingItemRatingsMock[0].noCriterionFulfilled}
      }
    },
    result: {
      data: {
        updateScenarioCodingItemRating: scenarioAutomatedCodingItemRatingsMock[0]
      }
    }
  }
]
