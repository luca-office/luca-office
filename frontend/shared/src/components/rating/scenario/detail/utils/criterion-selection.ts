import {isEqual} from "lodash-es"
import {AutomatedCodingCriterion, CodingCriterion, ScenarioCodingItemRating} from "../../../../../models"
import {exists, find, isAutomatedCodingCriterion, Option, sort} from "../../../../../utils"
import {ScenarioCriterionSelection} from "../../../models"

export const haveManualCriterionSelectionsChanged = (
  scenarioCodingItemRating: Option<ScenarioCodingItemRating>,
  criterionSelections: ScenarioCriterionSelection[]
): boolean =>
  scenarioCodingItemRating
    .map(
      rating =>
        !isEqual(
          sort(
            id => id ?? -1,
            rating.criterionSelections.map(selection => selection.manualCriterionId)
          ),
          sort(
            id => id ?? -1,
            criterionSelections.map(selection => selection.manualCriterionId)
          )
        )
    )
    .getOrElse(criterionSelections.length > 0)

export const haveAutomatedCriterionSelectionsChanged = (
  scenarioCodingItemRating: Option<ScenarioCodingItemRating>,
  criterionSelections: ScenarioCriterionSelection[]
): boolean =>
  scenarioCodingItemRating
    .map(
      rating =>
        !isEqual(
          sort(
            id => id ?? -1,
            rating.criterionSelections.map(selection => selection.automatedCriterionId)
          ),
          sort(
            id => id ?? -1,
            criterionSelections.map(selection => selection.automatedCriterionId)
          )
        )
    )
    .getOrElse(criterionSelections.length > 0)

export const getManualCriterionSelection = (
  criterionSelections: ScenarioCriterionSelection[],
  codingCriterion: CodingCriterion
): Option<ScenarioCriterionSelection> =>
  find(criterionSelection => criterionSelection.manualCriterionId === codingCriterion.id, criterionSelections)

export const getAutomatedCriterionSelection = (
  criterionSelections: ScenarioCriterionSelection[],
  codingCriterion: AutomatedCodingCriterion
): Option<ScenarioCriterionSelection> =>
  find(criterionSelection => criterionSelection.automatedCriterionId === codingCriterion.id, criterionSelections)

export const getCriterionSelections = (
  scenarioCodingItemRating: ScenarioCodingItemRating,
  codingCriteria: Array<AutomatedCodingCriterion | CodingCriterion>
): ScenarioCriterionSelection[] =>
  scenarioCodingItemRating.criterionSelections.filter(criterionSelection =>
    exists(
      codingCriterion =>
        isAutomatedCodingCriterion(codingCriterion)
          ? criterionSelection.automatedCriterionId === codingCriterion.id
          : criterionSelection.manualCriterionId === codingCriterion.id,
      codingCriteria
    )
  )
