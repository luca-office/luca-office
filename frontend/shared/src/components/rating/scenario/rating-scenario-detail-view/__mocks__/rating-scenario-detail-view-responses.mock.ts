import {pick, sumBy} from "lodash-es"
import {
  codingCriteriaMock,
  codingDimensionsMock,
  scenarioManualCodingItemRatingsMock
} from "../../../../../graphql/__mocks__"
import {ScenarioCodingItemRatingCreation} from "../../../../../graphql/generated/globalTypes"
import {createScenarioCodingItemRatingMutation} from "../../../../../graphql/mutations"
import {
  codingCriteriaQuery,
  codingDimensionsQuery,
  scenarioCodingItemRatingsQuery
} from "../../../../../graphql/queries"
import {ScenarioCodingItemRating} from "../../../../../models"
import {MockedResponse} from "@apollo/client/testing"

export const ratingScenarioDetailViewResponsesMock: MockedResponse[] = [
  {
    request: {
      query: codingDimensionsQuery,
      variables: {modelId: codingDimensionsMock[0].codingModelId}
    },
    result: {
      data: {
        codingDimensions: codingDimensionsMock.map((mock, index) => ({
          ...mock,
          parentDimensionId: index !== 1 ? codingDimensionsMock[1].id : mock.parentDimensionId,
          items: mock.items.map((item, index) => ({
            ...item,
            maximumScore: sumBy(codingCriteriaMock, ({score}) => score + index)
          }))
        }))
      }
    }
  },
  {
    request: {
      query: codingCriteriaQuery,
      variables: {itemId: codingDimensionsMock[0].items[0].id}
    },
    result: {
      data: {
        codingCriteria: codingCriteriaMock.map(mock => ({...mock, itemId: codingDimensionsMock[0].items[0].id}))
      }
    }
  },
  {
    request: {
      query: codingCriteriaQuery,
      variables: {itemId: codingDimensionsMock[2].items[0].id}
    },
    result: {
      data: {
        codingCriteria: codingCriteriaMock.map(mock => ({...mock, itemId: codingDimensionsMock[2].items[0].id}))
      }
    }
  },
  {
    request: {
      query: codingCriteriaQuery,
      variables: {itemId: codingDimensionsMock[2].items[1].id}
    },
    result: {
      data: {
        codingCriteria: codingCriteriaMock.map(mock => ({...mock, itemId: codingDimensionsMock[2].items[1].id}))
      }
    }
  },
  {
    request: {
      query: codingCriteriaQuery,
      variables: {itemId: codingDimensionsMock[2].items[2].id}
    },
    result: {
      data: {
        codingCriteria: codingCriteriaMock.map(mock => ({...mock, itemId: codingDimensionsMock[2].items[2].id}))
      }
    }
  },
  {
    request: {
      query: scenarioCodingItemRatingsQuery,
      variables: {ratingId: scenarioManualCodingItemRatingsMock[0].ratingId}
    },
    result: {
      data: {
        scenarioCodingItemRatings: scenarioManualCodingItemRatingsMock
      }
    }
  },
  {
    request: {
      query: createScenarioCodingItemRatingMutation,
      variables: {
        creation: pick<ScenarioCodingItemRating, keyof ScenarioCodingItemRatingCreation>(
          scenarioManualCodingItemRatingsMock[0],
          ["ratingId", "surveyInvitationId", "codingItemId"]
        )
      }
    },
    result: {
      data: {
        createScenarioCodingItemRating: scenarioManualCodingItemRatingsMock[0]
      }
    }
  }
]
