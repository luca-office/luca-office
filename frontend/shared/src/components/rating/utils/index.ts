export * from "./coding-criterion"
export * from "./coding-dimension"
export * from "./coding-item"
export * from "./question"
export * from "./score"
export * from "./scoring-type"
