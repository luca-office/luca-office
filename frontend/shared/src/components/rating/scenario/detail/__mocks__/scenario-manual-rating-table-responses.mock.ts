import {omit} from "lodash-es"
import {scenarioManualCodingItemRatingsMock} from "../../../../../graphql/__mocks__"
import {
  ScenarioCodingItemRatingCreation,
  ScenarioRatingCriterionSelectionCreation
} from "../../../../../graphql/generated/globalTypes"
import {
  createScenarioCodingItemRatingMutation,
  createScenarioRatingCriterionSelectionMutation,
  deleteScenarioRatingCriterionSelectionMutation,
  updateScenarioCodingItemRatingMutation
} from "../../../../../graphql/mutations"
import {ScenarioCodingItemRating, ScenarioRatingCriterionSelection} from "../../../../../models"
import {MockedResponse} from "@apollo/client/testing"

export const scenarioManualRatingTableResponsesMock: MockedResponse[] = [
  {
    request: {
      query: createScenarioCodingItemRatingMutation,
      variables: {
        creation: omit<
          ScenarioCodingItemRating,
          keyof Omit<ScenarioCodingItemRating, keyof ScenarioCodingItemRatingCreation>
        >(scenarioManualCodingItemRatingsMock[0], [
          "__typename",
          "id",
          "createdAt",
          "modifiedAt",
          "criterionSelections"
        ])
      }
    },
    result: {
      data: {
        createScenarioRating: scenarioManualCodingItemRatingsMock[0]
      }
    }
  },
  {
    request: {
      query: updateScenarioCodingItemRatingMutation,
      variables: {
        id: scenarioManualCodingItemRatingsMock[0].id,
        update: {noCriterionFulfilled: scenarioManualCodingItemRatingsMock[0].noCriterionFulfilled}
      }
    },
    result: {
      data: {
        updateScenarioCodingItemRating: scenarioManualCodingItemRatingsMock[0]
      }
    }
  },
  {
    request: {
      query: createScenarioRatingCriterionSelectionMutation,
      variables: {
        creation: omit<
          ScenarioRatingCriterionSelection,
          keyof Omit<ScenarioRatingCriterionSelection, keyof ScenarioRatingCriterionSelectionCreation>
        >(scenarioManualCodingItemRatingsMock[0].criterionSelections[0], ["__typename", "createdAt"])
      }
    },
    result: {
      data: {
        createScenarioRatingCriterionSelection: scenarioManualCodingItemRatingsMock[0].criterionSelections[0]
      }
    }
  },
  {
    request: {
      query: deleteScenarioRatingCriterionSelectionMutation,
      variables: {
        scenarioRatingId: scenarioManualCodingItemRatingsMock[0].id,
        manualCriterionId: scenarioManualCodingItemRatingsMock[0].criterionSelections[0].manualCriterionId
      }
    },
    result: {
      data: {
        deleteScenarioRatingCriterionSelection: scenarioManualCodingItemRatingsMock[0].criterionSelections[0]
      }
    }
  }
]
