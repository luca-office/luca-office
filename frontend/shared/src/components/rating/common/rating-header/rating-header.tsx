import {css} from "@emotion/react"
import * as React from "react"
import {ButtonConfig, ProjectModule} from "../../../../models"
import {CustomStyle, zIndex3} from "../../../../styles"
import {find, getProjectModuleIcon, getProjectModuleTitle, Option, sortByPosition} from "../../../../utils"
import {DetailViewHeader} from "../../../detail-view-header/detail-view-header"
import {CustomSelect} from "../../../select/custom-select"

export interface RatingHeaderProps extends CustomStyle {
  readonly navigationButtonConfig: ButtonConfig
  readonly onSelectModule: (projectModule: ProjectModule) => void
  readonly projectModules: ProjectModule[]
  readonly projectTitle: string
  readonly selectedModuleId: Option<UUID>
}

export const RatingHeader: React.FC<RatingHeaderProps> = ({
  customStyles,
  projectModules,
  projectTitle,
  navigationButtonConfig,
  onSelectModule,
  selectedModuleId
}) => {
  const dropdown = projectModules.length ? (
    <CustomSelect
      onChange={moduleId => find(module => module.id === moduleId, projectModules).forEach(onSelectModule)}
      optionList={sortByPosition(projectModules).map((module, index) => ({
        iconName: getProjectModuleIcon(module.moduleType),
        label: `${index + 1}. ${getProjectModuleTitle(module)}`,
        value: module.id
      }))}
      zIndexOfExpandedMenu={zIndex3}
      value={selectedModuleId.getOrElse("")}
      customStyles={styles.dropdown}
    />
  ) : undefined

  return (
    <DetailViewHeader
      customStyles={customStyles}
      labelKey={"rating__rating__header_label"}
      labelOptions={{title: projectTitle}}
      navigationButtonConfig={navigationButtonConfig}
      customContent={dropdown}
    />
  )
}

const Sizes = {
  dropdown: 386
}
const styles = {
  dropdown: css({
    marginLeft: "auto",
    width: Sizes.dropdown
  })
}
