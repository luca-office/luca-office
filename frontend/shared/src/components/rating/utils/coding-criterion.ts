import {meanBy, sumBy} from "lodash-es"
import {AutomatedCodingCriterion, CodingCriterion, ScenarioRatingCriterionSelection} from "../../../models"
import {exists, isAutomatedCodingCriterion} from "../../../utils"
import {ScenarioCriterionSelection} from "../models"

export const getScoreOfCodingCriteria = (codingCriteria: Array<CodingCriterion | AutomatedCodingCriterion>): number =>
  sumBy(codingCriteria, codingCriterion => codingCriterion.score)

// make sure criterionSelections is not empty, otherwise meanBy would return NaN
export const getAverageScoreOfCodingCriteria = (
  criterionSelections: Array<CodingCriterion | AutomatedCodingCriterion>
): number =>
  criterionSelections.length > 0 ? meanBy(criterionSelections, codingCriterion => codingCriterion.score) : 0

export const wasCodingCriterionRated = (
  isAutomatedCodingCriterion: boolean,
  codingCriterionId: UUID,
  scenarioRatingCriterionSelections: ScenarioRatingCriterionSelection[]
) =>
  exists(
    scenarioRatingCriterionSelection =>
      isAutomatedCodingCriterion
        ? scenarioRatingCriterionSelection.automatedCriterionId === codingCriterionId
        : scenarioRatingCriterionSelection.manualCriterionId === codingCriterionId,
    scenarioRatingCriterionSelections
  )

export const getSelectedCodingCriteria = (
  codingCriteria: Array<CodingCriterion | AutomatedCodingCriterion>,
  criterionSelections: Array<ScenarioCriterionSelection>
): Array<CodingCriterion | AutomatedCodingCriterion> =>
  codingCriteria.reduce<Array<CodingCriterion | AutomatedCodingCriterion>>(
    (accumulator, codingCriterion) =>
      exists(
        criterionSelection =>
          isAutomatedCodingCriterion(codingCriterion)
            ? criterionSelection.automatedCriterionId === codingCriterion.id
            : criterionSelection.manualCriterionId === codingCriterion.id,
        criterionSelections
      )
        ? [...accumulator, codingCriterion]
        : accumulator,
    []
  )
