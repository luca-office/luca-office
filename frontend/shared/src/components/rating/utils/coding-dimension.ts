import {CodingDimension, CodingItem} from "../../../models"
import {isDefined} from "../../../utils"

export const getCodingItemsFromCodingDimensions = (codingDimensions: CodingDimension[]): CodingItem[] =>
  codingDimensions.reduce(
    (accumulator, codingDimension) => [...accumulator, ...codingDimension.items],
    [] as CodingItem[]
  )

export const getSubDimensions = (
  codingDimension: CodingDimension,
  codingDimensions: CodingDimension[]
): CodingDimension[] =>
  isDefined(codingDimension.parentDimensionId)
    ? []
    : codingDimensions.filter(dimension => dimension.parentDimensionId === codingDimension.id)
