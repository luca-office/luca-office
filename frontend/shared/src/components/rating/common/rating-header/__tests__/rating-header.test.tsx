import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {projectModulesMockWithQuestionnaire} from "../../../../../graphql/__mocks__"
import {Option} from "../../../../../utils"
import {DetailViewHeader} from "../../../../detail-view-header/detail-view-header"
import {CustomSelect} from "../../../../select/custom-select"
import {RatingHeader, RatingHeaderProps} from "../rating-header"

const defaultProps: RatingHeaderProps = {
  projectModules: projectModulesMockWithQuestionnaire,
  navigationButtonConfig: {onClick: jest.fn, labelKey: "unit__days_plural"},
  onSelectModule: jest.fn(),
  projectTitle: "test title",
  selectedModuleId: Option.of(projectModulesMockWithQuestionnaire[1].id)
}

const getComponent = (props?: Partial<RatingHeaderProps>) => <RatingHeader {...defaultProps} {...props} />

describe("rating-header", () => {
  it("renders correctly", () => {
    const component = create(getComponent())

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", () => {
    const component = shallow(getComponent())
    expect(component.find(DetailViewHeader)).toHaveLength(1)
    expect(component.find(DetailViewHeader).prop("navigationButtonConfig")).toEqual(defaultProps.navigationButtonConfig)

    const select = component.find(DetailViewHeader).dive().find(CustomSelect)
    expect(select).toHaveLength(1)
    expect(select.prop("optionList")).toHaveLength(defaultProps.projectModules.length)
    expect(select.prop("value")).toEqual(defaultProps.selectedModuleId.getOrElse(""))
  })
})
