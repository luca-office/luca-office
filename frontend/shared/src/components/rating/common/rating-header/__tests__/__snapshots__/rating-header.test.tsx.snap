// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rating-header renders correctly 1`] = `
<div
  className="css-ul982y-DetailViewHeader"
>
  <button
    className="css-rmd9d0-Button"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="css-95j2hm-Button"
    >
      <div
        className="css-1lrzij5-Icon"
        dangerouslySetInnerHTML={
          {
            "__html": {},
          }
        }
      />
      <div
        className="css-u7npei-DetailViewHeader"
      >
        unit__days_plural
      </div>
    </div>
  </button>
  <div
    className="css-tqo6e0-DetailViewHeader"
  >
    rating__rating__header_label
  </div>
  <div
    className="css-19qjkf7"
    data-testid="custom-select"
  >
    <div
      className="css-0"
    />
    <div
      className="css-1vqpb7c"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="custom-select"
      >
        MockedSelect
      </div>
    </div>
  </div>
</div>
`;
